src/ReaderThemeSupportFeatures.php

```diff
@@ -0,0 +1,376 @@
+		);
+
+		// @todo Also do this when the theme has been updated?
+	}
+
+	/**
+	 * Check whether all the required props are present for a given feature item.
+	 *
+	 * @param string $feature Feature name.
+	 * @param array  $props   Props to check.
+	 *
+	 * @return bool Whether all are present.
+	 */
+	public function has_required_feature_props( $feature, $props ) {
+		if ( empty( $props ) || ! is_array( $props ) ) {
+			return false;
+		}
+		foreach ( self::SUPPORTED_FEATURES[ $feature ] as $required_prop ) {
+			if ( ! array_key_exists( $required_prop, $props ) ) {
+				return false;
+			}
+		}
+		return true;
+	}
+
+	/**
+	 * Filter the AMP options when they are updated to add the primary theme's features.
+	 *
+	 * @param array $options Options.
+	 * @return array Options.
+	 */
+	public function filter_amp_options_updating( $options ) {
+		if ( $this->reader_theme_loader->is_enabled( $options ) ) {
+			$options[ Option::PRIMARY_THEME_SUPPORT ] = $this->get_theme_support_features( true );
+		} else {
+			$options[ Option::PRIMARY_THEME_SUPPORT ] = null;
+		}
+		return $options;
+	}
+
+	/**
+	 * Update primary theme's cached theme support.
+	 */
+	public function update_cached_theme_support() {
+		if ( $this->reader_theme_loader->is_enabled() ) {
+			AMP_Options_Manager::update_option( Option::PRIMARY_THEME_SUPPORT, $this->get_theme_support_features( true ) );
+		} else {
+			AMP_Options_Manager::update_option( Option::PRIMARY_THEME_SUPPORT, null );
+		}
+	}
+
+	/**
+	 * Get the theme support features.
+	 *
+	 * @param bool $reduced Whether to reduce the feature props down to just what is required.
+	 * @return array Theme support features.
+	 */
+	private function get_theme_support_features( $reduced = false ) {
+		$features = [];
+		foreach ( array_keys( self::SUPPORTED_FEATURES ) as $feature_key ) {
+			$feature_value = current( (array) get_theme_support( $feature_key ) );
+			if ( ! is_array( $feature_value ) || empty( $feature_value ) ) {
+				continue;
+			}
+			if ( $reduced ) {
+				$features[ $feature_key ] = [];
+				foreach ( $feature_value as $item ) {
+					if ( $this->has_required_feature_props( $feature_key, $item ) ) {
+						$features[ $feature_key ][] = wp_array_slice_assoc( $item, self::SUPPORTED_FEATURES[ $feature_key ] );
+					}
+				}
+			} else {
+				$features[ $feature_key ] = $feature_value;
+			}
+		}
+		return $features;
+	}
+
+	/**
+	 * Add theme support from primary theme when a Reader theme has overridden.
+	 */
+	public function add_primary_theme_support() {
+		if ( ! $this->reader_theme_loader->is_theme_overridden() ) {
+			return;
+		}
+		$theme_support_features = AMP_Options_Manager::get_option( Option::PRIMARY_THEME_SUPPORT );
+		if ( ! is_array( $theme_support_features ) ) {
+			return;
+		}
+		foreach ( $theme_support_features as $support => $feature ) {
+			if ( is_array( $feature ) ) {
+				add_theme_support( $support, $feature ); // @todo Problem: $feature is now a subset.
```

Comment:
I thought this was a good idea at first, but now it may be problematic. Consider the case where a user has Twenty Thirteen as the primary theme and they select Twenty Twenty-One as the Reader theme. The TT1 theme actually reads from the `editor-color-palette` to populate a Customizer control for the background color, and the result here is that the color palette from Twenty Thirteen is showing up in the Customizer for Twenty Twenty-One:

![image](https://user-images.githubusercontent.com/134745/115639567-c9a39800-a2c9-11eb-8f64-4f76db441e4b.png)

On one hand that seems kinda cool, but it seems eventually it could cause problems, especially here in the case where I've reduced the theme support props to omit anything but the `slug` and the other prop that is needed to generate the styles. It could be that a theme wants to use the color values in the Customizer or some other UI as well, and they'd be missing here. 

I think we should consider not adding the theme support features from the primary theme. They're not actually used in the generation of the page styles since they're (currently) defined in the theme's `style.css`. But that could change in the future. There's nothing stopping someone from trying to be be more DRY and configurable and only define the color palette in PHP to then render out the style rules like we're doing here.

The bottom line is that we _are_ printing out the styles from the primary theme which are needed for rendering the blocks. Actually copying the theme support features from the primary theme would extend the scope to beyond the block styles to then affect other aspects of the theme design. As such we can revisit that later.